### PR TITLE
Make the cache key safe to use as a file name

### PR DESCRIPTION
**Pull request by guyradford**
| Question      | Answer
| ------------- | ------
| Bug fix?      | yes
| New feature?  |no
| BC breaks?    |no (only if cache not cleared)
| Deprecations? |no
| Tests pass?   | yes
| Fixed tickets | NA
| License       | MIT
| Doc PR        | reference to the documentation PR, if any

### Description

Some cache keys could not be used with the fileSystem adapter as they where not valid filenames. To solve this I base64 encoded the cache key and replaced = with _ to make a valid filename.

### TODO
* [ ] Add tests
* [ ] Add documentation
* [ X ] Updated Changelog.md

### CODE REVIEW

File: src/Adapter/Filesystem/FilesystemCachePool.php [hunk 1]
```diff
@@ -151,11 +151,10 @@ protected function storeItemInCache(CacheItemInterface $item, $ttl)
      */
     private function getFilePath($key)
     {
-        if (!preg_match('|^[a-zA-Z0-9_\.! ]+$|', $key)) {
```
Comment (aequasi) [inline src/Adapter/Filesystem/FilesystemCachePool.php]: I think this check should stay here...  I'm sure there are other characters that are not valid, that should be normalized as well, arent there?

**Comment by guyradford** [comment]
Shouldn't need it as base64 is only alpha numeric and = for padding. So
only the = will be a problem hence the replace.

**Comment by aequasi** [comment]
Yes, but we still don't want people creating keys with invalid characters, across the board.

**Comment by Nyholm** [comment]
What valid PSR6 characters does not work with the FileSystem adapter?

**Comment by guyradford** [comment]
@Nyholm lots, you are allowed pretty much any character at a minimum  `a-z` `A-Z` `0-9`  `.` ` _`. However the getFilePath was then validated against a smaller range `[a-zA-Z0-9_\.! ]` therefore disallowing some keys to be used as filenames. To fix this I propose to base64 encode the key, however this raised another issue for Windows users where case is ignored :( this would affect the current implementation too.

**Comment by guyradford** [comment]
As mentioned above currently there is an issue when this is used on a non-case sensitive OS like Windows, `CacheKey = cachekey`. Also base64 encoding also makes if difficult to locate the file, this is generally not an issue, but can be useful.

I was thinking that maybe to fix this we could apply the same translation that is used from CamelCase to underscores, where uppercase letters are proceeded by and underscore and the letter converted to lower. 

Any thoughts?

**Pull request closed without merging**